apps/ejabberd/src/ejabberd_c2s.hrl

```diff
@@ -0,0 +1,130 @@
+%% Directed presence unavailable remove user from pres_a.
+-record(state, {socket,
+                sockmod :: ejabberd:sockmod(),
+                socket_monitor,
+                xml_socket,
+                streamid,
+                sasl_state,
+                access,
+                shaper,
+                zlib = {false, 0}          :: {boolean(), integer()},
+                tls = false           :: boolean(),
+                tls_required = false  :: boolean(),
+                tls_enabled = false   :: boolean(),
+                tls_options = [],
+                authenticated = false :: boolean(),
+                jid                  :: ejabberd:jid(),
+                user = <<>>          :: ejabberd:user(),
+                server = <<>>     :: ejabberd:server(),
+                resource = <<>>      :: ejabberd:resource(),
+                sid                  :: ejabberd_sm:sid(),
+                %% We have _subscription to_ these users' presence status;
+                %% i.e. they send us presence updates.
+                %% This comes from the roster.
+                pres_t = ?SETS:new() :: gb_set(),
+                %% We have _subscription from_ these users,
+                %% i.e. they have subscription to us.
+                %% We send them presence updates.
+                %% This comes from the roster.
+                pres_f = ?SETS:new() :: gb_set(),
+                %% We're _available_ to these users,
+                %% i.e. we broadcast presence updates to them.
+                %% This may change throughout the session.
+                pres_a = ?SETS:new() :: gb_set(),
+                %% We are _invisible_ to these users.
+                %% This may change throughout the session.
+                pres_i = ?SETS:new() :: gb_set(),
+                pending_invitations = [],
+                pres_last, pres_pri,
+                pres_timestamp,
+                %% Are we invisible?
+                pres_invis = false :: boolean(),
+                privacy_list = #userlist{} :: mod_privacy:userlist(),
+                conn = unknown,
+                auth_module     :: ejabberd_auth:authmodule(),
+                ip              :: inet:ip_address(),
+                aux_fields = [] :: [{aux_key(), aux_value()}],
+                lang            :: ejabberd:lang(),
+                stream_mgmt = false,
+                stream_mgmt_in = 0,
+                stream_mgmt_id,
+                stream_mgmt_out_acked = 0,
+                stream_mgmt_buffer = [],
+                stream_mgmt_buffer_size = 0,
+                stream_mgmt_buffer_max = ?STREAM_MGMT_CACHE_MAX,
+                stream_mgmt_ack_freq = ?STREAM_MGMT_ACK_FREQ,
+                stream_mgmt_resume_timeout = ?STREAM_MGMT_RESUME_TIMEOUT,
+                stream_mgmt_resume_tref,
+                stream_mgmt_constraint_check_tref
+                }).
+-type aux_key() :: atom().
```

Comment:
I don't think putting types in .hrl files is a good idea due to possible name clashes. Namespacing options in Erlang are limited, but let's use them when possible ;)